master-docker-nonstandard/master.cfg

```diff
@@ -793,10 +793,43 @@ f_valgrind_build.addStep(getSourceTarball())
+                """cmake . \\
+                    -DCMAKE_C_COMPILER_LAUNCHER=ccache \\
+                    -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \\
+                    -DCMAKE_C_FLAGS="-Wall -Wextra -Wunused -Wwrite-strings -Wno-uninitialized -Wno-strict-aliasing -Wformat-security -Wvla -Wimplicit-fallthrough=2 -mtune=native -m64 -DEXTRA_DEBUG -DSAFE_MUTEX -DSAFEMALLOC -O0 -g3 -gdwarf-2 -DFORCE_INIT_OF_VARS -Wuninitialized -DHAVE_valgrind -USAFEMALLOC -DTRASH_FREE_MEMORY -UFORCE_INIT_OF_VARS -Wno-uninitialized -DMYSQL_SERVER_SUFFIX=-valgrind-max" \\
+                    -DCMAKE_CXX_FLAGS="-Wall -Wextra -Wunused -Wwrite-strings -Wno-uninitialized -Wno-strict-aliasing -Wformat-security -Wvla -Wno-unused-parameter -Wno-invalid-offsetof -Wnon-virtual-dtor -Wimplicit-fallthrough=2 -felide-constructors -fexceptions -mtune=native -m64 -DEXTRA_DEBUG -DSAFE_MUTEX -DSAFEMALLOC -O0 -g3 -gdwarf-2 -DFORCE_INIT_OF_VARS -Wuninitialized -DHAVE_valgrind -USAFEMALLOC -DTRASH_FREE_MEMORY -UFORCE_INIT_OF_VARS -Wno-uninitialized -DMYSQL_SERVER_SUFFIX=-valgrind-max" \\
+                    -DENABLE_ASSEMBLER=1 \\
+                    -DWITH_EXTRA_CHARSETS=complex \\
+                    -DENABLE_THREAD_SAFE_CLIENT=1 \\
+                    -DWITH_BIG_TABLES=1 \\
+                    -DWITH_PLUGIN_ARIA=1 \\
+                    -DWITH_ARIA_TMP_TABLES=1 \\
+                    -DWITH_JEMALLOC=NO \\
+                    -DCMAKE_BUILD_TYPE=Debug \\
+                    -DSECURITY_HARDENED=OFF \\
+                    -DWITH_VALGRIND=1 \\
+                    -DHAVE_LIBAIO_H=0 \\
+                    -DCMAKE_DISABLE_FIND_PACKAGE_URING=1 \\
+                    -DCMAKE_DISABLE_FIND_PACKAGE_LIBAIO=1 \\
+                    -DWITH_SSL=system \\
+                    -DWITH_MAX=AUTO \\
+                    -DWITH_EMBEDDED_SERVER=1 \\
+                    -DWITH_LIBEVENT=bundled \\
+                    -DPLUGIN_PLUGIN_FILE_KEY_MANAGEMENT=NO \\
+                    -DPLUGIN_TEST_SQL_DISCOVERY=DYNAMIC \\
+                    -DPLUGIN_TOKUDB=NO \\
+                    -DPLUGIN_ROCKSDB=DYNAMIC \\
+                    -DPLUGIN_AUTH_GSSAPI=DYNAMIC \\
+                    -DENABLE_LOCAL_INFILE=1 \\
+                    -DMYSQL_SERVER_SUFFIX=-valgrind-max \\
+                    -DWITH_DBUG_TRACE=OFF \\
+                    -DPLUGIN_S3=STATIC \\
+                    -DPLUGIN_FILE_KEY_MANAGEMENT=DYNAMIC \\
+                    -DPLUGIN_HASHICORP_KEY_MANAGEMENT=DYNAMIC \\
```

Comment:
Sorry for taking me so long to respond.  I took some time to investigate and digest those written by you. Also thank you very much for such a detailed review which opened new territories for me to explore.

I want to confirm a few things with you, case by case.

**[1] DWITH_BIG_TABLES** 

Variable doesn’t seem do to anything if provided to CMAKE. I don’t see it in any CMAKE invokation, and moreover, BIG_TABLES it’s defined under a block of macros which states that compilation will break without them.

https://github.com/MariaDB/server/blob/7c5fdc9b6a45f4e8bad1ae5c7c93b457be9891fa/config.h.cmake#L510


**[2] DHAVE_LIBAIO_H / DCMAKE_DISABLE_FIND_PACKAGE_URING / DCMAKE_DISABLE_FIND_PACKAGE_LIBAI**O

When running without setting these to OFF, I’d expect that LIBAIO and URING will be used, instead from the –trace-expand I get that they are optional by default.

    • /buildbot/tpool/CMakeLists.txt(5):  OPTION(WITH_URING Require that io_uring be used OFF )
    • /buildbot/tpool/CMakeLists.txt(6):  OPTION(WITH_LIBAIO Require that libaio is used, unless uring is there OFF ).

If I run with -DWITH_LIBAIO it won’t check the existence of the package because there is a typo in the code, here and also if URING is present, the code block for checking and applying settings when LIBAIO is present, won’t run.

https://github.com/MariaDB/server/blob/7c5fdc9b6a45f4e8bad1ae5c7c93b457be9891fa/tpool/CMakeLists.txt#L10

ELSEIF(WITH_LIBAIO)
SET(LIBAIO_REQIRED REQUIRED)

ELSE()
FIND_PACKAGE(LIBAIO QUIET ${LIBAIO_REQUIRED})

Anyway, since LIBURING is present on the image it will still apply the settings under IF(FOUND_URING).

**[3] FORCE_INIT_OF_VARS**
- Not sure if I understood correctly if this macro should be defined or not for a Valgrind build.
Looked in the server code and it seems that the default is undefined, the only exception being the connect storage engine.



**[4]  WITH_LIBEVENT**

Yes, it is part of the container image.
libevent_pthreads-2.1.so.7 (libc6,x86-64) => /lib/x86_64-linux-gnu/libevent_pthreads-2.1.so.7
	libevent_openssl-2.1.so.7 (libc6,x86-64) => /lib/x86_64-linux-gnu/libevent_openssl-2.1.so.7
	libevent_extra-2.1.so.7 (libc6,x86-64) => /lib/x86_64-linux-gnu/libevent_extra-2.1.so.7
	libevent_core-2.1.so.7 (libc6,x86-64) => /lib/x86_64-linux-gnu/libevent_core-2.1.so.7
	libevent-2.1.so.7 (libc6,x86-64) => /lib/x86_64-linux-gnu/libevent-2.1.so.7

**[5] – MYSQL_SERVER_SUFFIX**

I discovered some MTR tests that depends on the version if is set to Valgrind.

https://github.com/MariaDB/server/blob/7c5fdc9b6a45f4e8bad1ae5c7c93b457be9891fa/mysql-test/include/not_valgrind.inc

But I suppose that even if not setting the SERVER_SUFFIX, “Valigrind” will still be appended to the version (based on the line below) and tests will work just fine.

https://github.com/MariaDB/server/blob/7c5fdc9b6a45f4e8bad1ae5c7c93b457be9891fa/sql/mysqld.cc#L8953

**[6]: PLUGIN_TOKUDB = NO**

What is the rationale of this being disabled for Valgrind builds?
Also, it will override WITH_MAX=ON ?

**[7] WITH_SSL**

If I understand correctly the default is “system” so it will invoke Find_Package in CMAKE to satisfy this requirement.

On the Valgrind image found:
	libssl-dev/jammy-updates,jammy-security 3.0.2-0ubuntu1.16 amd64 [upgradable from: 3.0.2-0ubuntu1.7]

Reply:
[1] correct
[2] well spotted - https://github.com/MariaDB/server/pull/3375 , 10.5 still depends on libaio, 10.6+ will be uring.
[3] correct
[4] don't forget devel package needed for include headers - libevent shared libraries are insufficient
[5] `$VALGRIND_TEST` is set by the mtr option `--valgrind` of [mtr](https://github.com/MariaDB/server/blob/7c5fdc9b6a45f4e8bad1ae5c7c93b457be9891fa/mysql-test/mariadb-test-run.pl#L5164)
[6] turn out this is disabled by default. - effectively unmaintained - https://mariadb.com/kb/en/tokudb/
[7] yes

Valgrind image - consider a bump to nobel [3.22](https://packages.ubuntu.com/noble/valgrind), or fedora40 (3.23).  The [3.23](https://valgrind.org/docs/manual/dist.news.html) might be required for `__builtin_strcmp` (or ` -fno-builtin-strcmp`, or the current msse hack, which is counter intuitive to testing as close to production.)

Reply:
For [4] - totally missed that line when looking through the package list. Yes, it is on the current image:

_- libevent-dev/jammy,now 2.1.12-stable-1build3 amd64 [installed]_

Reply:
for [6] - I guess I can for sure omit the option PLUGIN_TOKUDB = NO
as every storage engine CMakeLists.txt is calling on MYSQL_ADD_PLUGIN . As TOKUDB is missing, we should be fine.